src/v2.rs

```diff
@@ -169,12 +178,141 @@ impl EszipV2Modules {
   }
 }
 
+#[derive(Debug, Clone, Copy)]
+struct Options {
+  /// Hash Function used to checksum the contents of the eszip when encoding/decoding
+  ///
+  /// If the eszip does not include the option, it defaults to `[Checksum::NoChecksum]` in >=v2.2
+  /// and `[Checksum::Sha256]` in older versions.  It is `None` when the eszip header includes a
+  /// checksum that this version of the library does not know.
+  checksum: Option<Checksum>,
+
+  /// Size in Bytes of the hash function digest.
+  ///
+  /// Defaults to the known length of the configured hash function. Useful in order to ensure forwards compatibility,
+  /// otherwise the parser does not know how many bytes to read.
+  checksum_size: Option<u8>,
+}
+
+impl Options {
+  fn default_for_version(magic: &[u8; 8]) -> Self {
+    let defaults = Self {
+      checksum: Some(Checksum::NoChecksum),
+      checksum_size: Default::default(),
+    };
+    #[cfg(feature = "sha256")]
+    let mut defaults = defaults;
+    if let ESZIP_V2_MAGIC | ESZIP_V2_1_MAGIC = magic {
+      // versions prior to v2.2 default to checksuming with SHA256
+      #[cfg(feature = "sha256")]
+      {
+        defaults.checksum = Some(Checksum::Sha256);
+      }
+    }
+    defaults
+  }
+}
+
+impl Default for Options {
+  fn default() -> Self {
+    Self::default_for_version(LATEST_VERSION)
+  }
+}
+
+impl Options {
+  /// Get the size in Bytes of the source hashes
+  ///
+  /// If the eszip has an explicit digest size, returns that. Otherwise, returns
+  /// the default digest size of the [`Self::checksum`]. If the eszip
+  /// does not have either, returns `None`.
+  fn checksum_size(self) -> Option<u8> {
+    self
+      .checksum_size
+      .or_else(|| Some(self.checksum?.digest_size()))
+  }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u8)]
+pub enum Checksum {
+  NoChecksum = 0,
+  #[cfg(feature = "sha256")]
+  Sha256 = 1,
+  #[cfg(feature = "crc32")]
+  Crc32 = 2,
+  #[cfg(feature = "xxhash")]
+  XxHash = 3,
+  #[cfg(feature = "xxhash3")]
+  XxHash3 = 4,
+}
+
+impl Checksum {
+  const fn digest_size(self) -> u8 {
+    match self {
+      Self::NoChecksum => 0,
+      #[cfg(feature = "sha256")]
+      Self::Sha256 => 32,
+      #[cfg(feature = "crc32")]
+      Self::Crc32 => 4,
+      #[cfg(feature = "xxhash")]
+      Self::XxHash => 8,
+      #[cfg(feature = "xxhash3")]
+      Self::XxHash3 => 8,
+    }
+  }
+
+  fn from_u8(discriminant: u8) -> Option<Self> {
+    Some(match discriminant {
+      0 => Self::NoChecksum,
+      #[cfg(feature = "sha256")]
+      1 => Self::Sha256,
+      #[cfg(feature = "crc32")]
+      2 => Self::Crc32,
+      #[cfg(feature = "xxhash")]
+      3 => Self::XxHash,
+      #[cfg(feature = "xxhash3")]
+      4 => Self::XxHash3,
+      _ => return None,
```

Comment:
Perhaps this should error when it finds something it doesn't recognize? Otherwise it might lead to corrupt data being loaded I think.

Reply:
The idea is to have a forward-compatible design that falls back to no checksuming when encountering an unknown hash function. This applies both to decoding an eszip built with a newer version of the library, and built with a hash function not enabled for your library. 

This does not lead to corrupt data thanks to being explicit about the checksum size in the options header. Even if the library does not know how to validate it, it still knows how long the checksum is. Users are expected to use [`EszipV2::is_checksumed()` and `EszipV2::should_be_checksumed()`](https://github.com/denoland/eszip/blob/fa49cb5fa4622b590270976d38855081daaa1f53/src/v2.rs#L744-L762) to ensure the content has been checksumed, if they rely on this guarantee. 

See https://github.com/denoland/eszip/blob/fa49cb5fa4622b590270976d38855081daaa1f53/src/v2.rs#L2444-L2482